### PR TITLE
Refactor MediaTypedEncoder/MediaTypedDecoder

### PR DESCRIPTION
**Pull request by larioj**
* The refactor will make MediaTypedEncoder and
  MediaTypedDecoder consistent with each other
  * We want to leverage these types to encode and decode things that are not http requests.
  * Move MediaTypedDecoder to
    MediaTypedRequest
  * Create MediaTypedDecoder that is the
    Decoder version of MediaTypedEncoder

### CODE REVIEW

File: cosmos-server/src/main/scala/com/mesosphere/cosmos/rpc/v1/circe/MediaTypedDecoders.scala [hunk 1]
```diff
@@ -1,44 +1,44 @@
 package com.mesosphere.cosmos.rpc.v1.circe
 
-import com.mesosphere.cosmos.circe.MediaTypedDecoder
+import com.mesosphere.cosmos.circe.MediaTypedRequestDecoder
 import com.mesosphere.cosmos.http.MediaTypes
 import com.mesosphere.cosmos.rpc.v1.circe.Decoders._
 import com.mesosphere.cosmos.rpc.v1.model._
 import io.finch.circe._
 
 object MediaTypedDecoders {
```
Comment (jsancio) [inline cosmos-server/src/main/scala/com/mesosphere/cosmos/rpc/v1/circe/MediaTypedDecoders.scala]: Please rename the file and object to `MediaTypedRequestDecoder`.

File: cosmos-server/src/main/scala/com/mesosphere/cosmos/circe/MediaTypedRequestDecoder.scala [hunk 2]
```diff
@@ -0,0 +1,32 @@
+package com.mesosphere.cosmos.circe
```
Comment (jsancio) [inline cosmos-server/src/main/scala/com/mesosphere/cosmos/circe/MediaTypedRequestDecoder.scala]: `DecodeRequest` comes from `io.finch` we should change this filename and package to `com.mesosphere.cosmos.finch`.

We talked about not needing this class at all and instead using `MediaTypedDecoder`. What made you change from that plan? Were you not able to do this? Please explain why. Thanks!

**Comment by larioj** [comment]
@jsancio The change that we discussed did not work. It broke six unit tests. I didn't see much value into delving any further given that the difference in quality between the solutions is small

File: cosmos-server/src/main/scala/com/mesosphere/cosmos/circe/MediaTypedEncoder.scala [hunk 3]
```diff
@@ -8,6 +8,12 @@ final class MediaTypedEncoder[A] private(val encoder: Encoder[A], val mediaType:
 
 object MediaTypedEncoder {
 
+  def apply[A](mediaType: MediaType)(implicit 
+    encoder: Encoder[A]
+  ): MediaTypedEncoder[A] = {
+    MediaTypedEncoder(encoder, mediaType)
+  }
```
Comment (BenWhitehead) [inline cosmos-server/src/main/scala/com/mesosphere/cosmos/circe/MediaTypedEncoder.scala]: This method doesn't appear to be used, why was it added?

**Comment by BenWhitehead** [comment]
Based on [this](https://github.com/dcos/cosmos/pull/482#r78452261) comment in #482 I expected something more like:

```scala
case class MediaTypedDecoder[A](decoder: io.circe.Decoder[A], mediaType: MediaType)
case class MediaTypedResponseDecoder[A](mtd: MediaTypedDecoder[A]) {
  val finchDecoder: io.finch.DecoderRequest[A] = ???
}
```

Why did you end up with the duplication between MediaTypedDecoder and MediaTypedResponseDecoder?

**Comment by larioj** [comment]
* The purpose of this refactoring is to utilize
  the MediaTypedCoders in the serialization of
  classes into storage. Currently they are being
  used only to serialize classes for http requests
  and responses
* Move MediaTypedDecoder to MediaTypedRequestDecoder
  * MediaTypedRequestDecoder is a distinct type from
    MediaTypedDecoder. MediaTypedDecoder associates
    a circe.Decoder with a mediaType while
    MediaTypedRequestDecoder associates a
    finch.DecodeRequest with other metadata.
  * MediaTypedRequestDecoder will use MediaTypedDecoder
    to avoid duplication of work.
* Create MediaTypedDecoder that is the
    Decoder version of MediaTypedEncoder
* The refactor will make MediaTypedEncoder and
  MediaTypedDecoder consistent with each other
  * MediaTypedEncoder will get a method that takes
    parameters implicitly. This is not currently used
    in the code base, but will make the encoder usage
    similar to the decoder usage moving forward.
* Will fix unit tests that currently assume we decoding
  of things other the Json. This change will make it so
  that the MediaTypedCoders can only handle Json

File: cosmos-server/src/main/scala/com/mesosphere/cosmos/circe/MediaTypedEncoder.scala (hunk 3 above)
Comment (larioj) [inline cosmos-server/src/main/scala/com/mesosphere/cosmos/circe/MediaTypedEncoder.scala]: It makes this class symmetric with MediaTypedDecoder

**Comment by larioj** [comment]
Closing this PR. This PR should be made to master rather than distributed publish. See the new PR #499

**Pull request closed without merging**